flareon/src/request.rs

```diff
@@ -0,0 +1,97 @@
+use std::borrow::Cow;
+use std::sync::Arc;
+
+use bytes::Bytes;
+use indexmap::IndexMap;
+
+use crate::{Error, FlareonProject, FORM_CONTENT_TYPE};
+
+#[derive(Debug)]
+pub struct Request {
+    inner: axum::extract::Request,
+    project: Arc<FlareonProject>,
+    pub(crate) path_params: IndexMap<String, String>,
+}
+
+impl Request {
+    #[must_use]
+    pub fn new(inner: axum::extract::Request, project: Arc<FlareonProject>) -> Self {
+        Self {
+            inner,
+            project,
+            path_params: IndexMap::new(),
+        }
+    }
+
+    #[must_use]
+    pub fn inner(&self) -> &axum::extract::Request {
+        &self.inner
+    }
+
+    #[must_use]
+    pub fn project(&self) -> &FlareonProject {
+        &self.project
+    }
+
+    #[must_use]
+    pub fn uri(&self) -> &axum::http::Uri {
+        self.inner.uri()
+    }
+
+    #[must_use]
+    pub fn method(&self) -> &axum::http::Method {
+        self.inner.method()
+    }
+
+    #[must_use]
+    pub fn headers(&self) -> &axum::http::HeaderMap {
+        self.inner.headers()
+    }
+
+    #[must_use]
+    pub fn content_type(&self) -> Option<&axum::http::HeaderValue> {
+        self.inner.headers().get(axum::http::header::CONTENT_TYPE)
+    }
+
+    pub async fn form_data(&mut self) -> Result<Bytes, Error> {
```

Comment:
Perhaps. But then won't it be confusing when we want to have a method that returns a raw request body? `data()` sounds very much like it could be just returning raw data, while this method specifically parses the URL/request body as a HTML form, hence the name.